### ISSUE TITLE
kv: pebbleReadOnly constructed in replicaRaftStorage.Entries even when reading from entry cache

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#92274]
I noticed in a recent round of benchmarking that we construct a `pebbleReadOnly` in `replicaRaftStorage.Entries` even in the vastly common case where all entries are served out of the raft entry cache:

https://github.com/cockroachdb/cockroach/blob/2a041d61d021f563828597fe9961f54727419c1d/pkg/kv/kvserver/replica_raftstorage.go#L102

This is wasteful.

I would have made the fix of pushing this into `entries` myself, but there's was just enough friction due to `term`'s [use of the `reader`](https://github.com/cockroachdb/cockroach/blob/2a041d61d021f563828597fe9961f54727419c1d/pkg/kv/kvserver/replica_raftstorage.go#L296) that I figured I'd leave it to the experts.

Is it worth asking whether this code needs to operate on a LSM snapshot? There should not be concurrent mutation to the state it is looking at.

**Comment by erikgrinaker** [comment]
@tbg @pavelkalinnikov This seems like something we could easily address with the ongoing Raft log storage work.

**Comment by tbg** [comment]
Sounds good, added to CRDB-220.

**Comment by pavelkalinnikov** [comment]
Worth noting that this construction doesn't look particularly heavy:
https://github.com/cockroachdb/cockroach/blob/2ace40eba5e9a9423231d7a73dfa6db0c0128263/pkg/storage/pebble.go#L1887-L1903

But in principle, it makes sense to reverse the order.

**Comment by erikgrinaker** [comment]
Yeah, good point -- this isn't taking out an engine snapshot immediately, it's pinning the iterators on first access, reusing existing pooled iterators. This is probably cheaper than constructing a new iterator straight from the engine when we do end up using it though.

@nvanbenschoten Since you bring this up, I assume construction of `pebbleReadOnly` showed up prominently enough in these benchmarks to make it worth optimizing, even though the construction cost is just fetching a pooled struct and reinitializing it?

**Comment by nvanbenschoten** [comment]
It wasn't prominent in these benchmarks, but it did jump out when looking at a CPU profile of kv0 and focusing on Raft. The pebbleReadOnly construction accounted for about **0.1%** of the total CPU profile. However, Raft itself was only about **10%**, so you can back into this being around a **1%** savings in the compute work performed by Raft.

Here's the one screenshot I captured of this (a week or so ago).

<img width="1756" alt="Screen Shot 2022-11-11 at 2 24 29 AM" src="https://user-images.githubusercontent.com/5438456/203392092-10ab75d8-edc8-42a2-915e-1baf4c13fa71.png">

That's not a huge saving, but it is still probably worth doing while we're working in this area.

I think there's also something nice about not touching the engine until we need to.
